spark/src/test/scala/org/apache/spark/sql/delta/DeltaProtocolVersionSuite.scala

```diff
@@ -420,6 +420,185 @@ trait DeltaProtocolVersionSuiteBase extends QueryTest
     }
   }
 
+  test("InvalidProtocolVersionException - error message with protocol too high - table path") {
+    withTempDir { path =>
+      spark.range(1).write.format("delta").save(path.getCanonicalPath)
+      val (deltaLog, snapshot) = DeltaLog.forTableWithSnapshot(spark, path.getCanonicalPath)
+
+      var tableReaderVersion = 4
+      var tableWriterVersion = 7
+      var version = snapshot.version + 1
+      untrackedChangeProtocolVersion(deltaLog, version, tableReaderVersion, tableWriterVersion)
+
+      val exceptionRead = intercept[InvalidProtocolVersionException] {
+        spark.read.format("delta").load(path.getCanonicalPath)
+      }
+      assert(exceptionRead.getMessage == getExpectedProtocolErrorMessage(
+        deltaLog.dataPath.toString,
+        tableReaderVersion,
+        tableWriterVersion))
+
+      tableReaderVersion = 3
+      tableWriterVersion = 8
+      version = version + 1
+      untrackedChangeProtocolVersion(deltaLog, version, tableReaderVersion, tableWriterVersion)
+
+      val exceptionWrite = intercept[InvalidProtocolVersionException] {
+        spark.range(1).write
+          .mode("append")
+          .option("mergeSchema", "true")
+          .format("delta")
+          .save(path.getCanonicalPath)
+      }
+      assert(exceptionWrite.getMessage == getExpectedProtocolErrorMessage(
+        deltaLog.dataPath.toString,
+        tableReaderVersion,
+        tableWriterVersion))
+    }
+  }
+
+  def testInvalidProtocolErrorMessageWithTableName(warm: Boolean): Unit = {
+    val protocolTableName = "mytableprotocoltoohigh"
+    withTable(protocolTableName) {
+      spark.range(1).write.format("delta").saveAsTable(protocolTableName)
+      val (deltaLog, snapshot) = DeltaLog.forTableWithSnapshot(
+        spark,
+        TableIdentifier(protocolTableName))
+
+      var tableReaderVersion = 4
+      var tableWriterVersion = 7
+      var version = snapshot.version + 1
+      untrackedChangeProtocolVersion(deltaLog, version, tableReaderVersion, tableWriterVersion)
+      if (!warm) {
+        DeltaLog.clearCache()
+      }
+
+      val exceptionRead = intercept[InvalidProtocolVersionException] {
+        spark.read.format("delta").table(protocolTableName)
+      }
+
+      var pathInErrorMessage = "default." + protocolTableName
+      val readErrorMessage = getExpectedProtocolErrorMessage(
+        pathInErrorMessage,
+        tableReaderVersion,
+        tableWriterVersion)
+
+      assert(exceptionRead.getMessage == readErrorMessage)
+
+      val exceptionMerge = intercept[InvalidProtocolVersionException] {
+        val deltaTable = io.delta.tables.DeltaTable.forName(protocolTableName)
+        deltaTable.merge(spark.range(1).as("src").toDF, s"src.id = $protocolTableName.id")
+          .whenMatched()
+          .updateAll()
+          .execute()
```

Comment:
nit tl;dr: The merge test probably doesn't add much value in its current form, but is also completely harmless so not a big deal either way.

I'm not sure it's useful to have the merge inside the `intercept` right now, because `DeltaTable` always creates an internal dataframe, which requires a schema, which will trigger a snapshot update even for warm case?

If we're specifically trying to test `merge`, we will need to create a "stale" DeltaTable before invalidating the table's protocol, so the merge internally triggers the `DeltaLog.update` call.

If we're trying to test the `DeltaTable.forName`, should probably put it in its own `intercept`?

But at some point, I'm not sure this test case is even necessary any more, because the PR now has a very narrow waist? The merge must trigger a transaction at some point internally, and when it does, we know what will happen because we already test the write path in isolation.

Reply:
I think makes sense to remove it. I added when I was running experiments, but it is not testing any different code path.